StyleCop.Analyzers/StyleCop.Analyzers/GeneratedCodeAnalysisExtensions.cs

```diff
@@ -0,0 +1,108 @@
+
+        internal static bool IsGenerated(this SyntaxTreeAnalysisContext context, CancellationToken cancellationToken) => context.Tree?.IsGenerated(cancellationToken) ?? false;
+
+        internal static bool IsGenerated(this SymbolAnalysisContext context, CancellationToken cancellationToken)
+        {
+            if (context.Symbol == null) return false;
+            foreach (var syntaxReference in context.Symbol.DeclaringSyntaxReferences)
+            {
+                if (syntaxReference.SyntaxTree.IsGenerated(cancellationToken)) return true;
+                var root = syntaxReference.SyntaxTree.GetRoot(cancellationToken);
+                var node = root?.FindNode(syntaxReference.Span);
+                if (node.IsGenerated()) return true;
+            }
+            return false;
+        }
+
+        internal static bool IsGenerated(this SyntaxNode node) => node.HasAttributeOnAncestorOrSelf("DebuggerNonUserCode", "GeneratedCode");
+
+        internal static bool IsGenerated(this SyntaxTree tree, CancellationToken cancellationToken) => (tree.FilePath?.IsOnGeneratedFile() ?? false) || tree.HasAutoGeneratedComment(cancellationToken);
+
+        internal static bool HasAutoGeneratedComment(this SyntaxTree tree, CancellationToken cancellationToken)
+        {
+            lock (GeneratedHeaderPresentCheck)
+            {
+                object previouslyDeterminedValue;
+                if (GeneratedHeaderPresentCheck.TryGetValue(tree, out previouslyDeterminedValue))
+                {
+                    return previouslyDeterminedValue == GeneratedHeaderPresent;
+                }
+            }
+
+            var root = tree.GetRoot(cancellationToken);
+
+            if (root == null) return false;
+            var firstToken = root.GetFirstToken();
+            SyntaxTriviaList trivia;
+            if (firstToken == default(SyntaxToken))
+            {
+                var token = ((CompilationUnitSyntax)root).EndOfFileToken;
+                if (!token.HasLeadingTrivia) return false;
+                trivia = token.LeadingTrivia;
+            }
+            else
+            {
+                if (!firstToken.HasLeadingTrivia) return false;
+                trivia = firstToken.LeadingTrivia;
+            }
+
+            var commentLines = trivia.Where(t => t.IsKind(SyntaxKind.SingleLineCommentTrivia)).Take(2).ToList();
+            if (commentLines.Count != 2) return false;
+            bool autoGenerated = commentLines[1].ToString() == "// <auto-generated>";
+
+            lock (GeneratedHeaderPresentCheck)
+            {
+                // We don't know if there was a race, and we want to avoid throwing,
+                // so now that we have the lock, remove the key if it is already there.
+                GeneratedHeaderPresentCheck.Remove(tree);
+                GeneratedHeaderPresentCheck.Add(tree, autoGenerated ? GeneratedHeaderPresent : GeneratedHeaderAbsent);
+            }
```

Comment:
Yup. That makes sense since ConditionalWeakTable is thread-safe and its entry is now guaranteed. Nice.